### PR TITLE
feat(topology): adds support for decorator in topology for ingress

### PR DESCRIPTION
**Pull request by invincibleJai**
**Tracks:** https://github.com/janus-idp/backstage-plugins/issues/58

**Fixes:**  https://github.com/janus-idp/backstage-plugins/issues/123

**Screenshots/GIF's:**
- Light Theme
image

- Dark Theme
image

**Test Setup:**
Follow https://github.com/janus-idp/backstage-plugins/tree/main/plugins/topology#readme

**Unit Tests:**
image

**Todo:**
- [x] adds support for decorator in topology for ingress
- [x] adds accessibility styles from pf (for control bars)
- [x] add unit tests

### CODE REVIEW

**Review by jeff-phillips-18 — APPROVED** [review]

**Comment by invincibleJai** [comment]
Thanks @jeff-phillips-18 for the review

@divyanshiGupta @spadgett Needed your help with review, PTAL.

**Review by spadgett — COMMENTED** [review]
Thanks, @invincibleJai. I'm not super familiar with the ingress API, but I had a few questions on the logic. We might not be handling some edge cases.

File: plugins/topology/src/utils/label-selector.js [hunk 1]
```diff
@@ -0,0 +1,285 @@
+/* eslint-disable guard-for-in, default-case, no-case-declarations */
+
+/**
+ *    Copied from "kubernetes-label-selector"
```
[the remark this answers is not in the document]
Comment (invincibleJai) [reply plugins/topology/src/utils/label-selector.js:4]: Yeah Sam,  it's taken from the console

File: plugins/topology/src/utils/resource-utils.ts [hunk 2]
```diff
@@ -42,3 +49,91 @@ export const createOverviewItemForType = (
...
+  return services.filter((service: V1Service) => {
+    const specSelector = service.spec?.selector || {};
+    const selector = new LabelSelector(specSelector);
+    return selector.matches(template);
+  });
+};
+
+export const getIngressWebURL = (ingress: V1Ingress): string | null => {
+  const scheme = ingress.spec?.tls ? 'https' : 'http';
+  const { host, http } = ingress.spec?.rules?.[0] || {};
```
[the remark this answers is not in the document]
Comment (invincibleJai) [reply plugins/topology/src/utils/resource-utils.ts]: yeah updated it as an ingress can have rules associated with multiple services

File: plugins/topology/src/utils/resource-utils.ts [hunk 3]
```diff
@@ -42,3 +49,91 @@ export const createOverviewItemForType = (
...
+  if (!host) {
+    return null;
+  }
+  let url = `${scheme}://${host}`;
```
[the remark this answers is not in the document]
Comment (invincibleJai) [reply plugins/topology/src/utils/resource-utils.ts]: yeah have handled it, route decorator will be shown only if the host exists and if there are now wildcards. In sidebar we can show these details/

**Review by spadgett — APPROVED** [review]
lgtm

**Pull request merged**